pkg/reconciler/taskrun/resources/taskspec.go

```diff
@@ -105,12 +99,13 @@ func GetTaskData(ctx context.Context, taskRun *v1.TaskRun, getTask GetTask, getS
 }
 
 // extractStepActions extracts the StepActions and merges them with the inlined Step specification.
-func extractStepActions(ctx context.Context, taskSpec v1.TaskSpec, getStepAction GetStepAction) ([]v1.Step, error) {
+func GetStepActionsData(ctx context.Context, taskSpec v1.TaskSpec, tr *v1.TaskRun, tekton clientset.Interface) ([]v1.Step, error) {
 	steps := []v1.Step{}
 	for _, step := range taskSpec.Steps {
 		s := step.DeepCopy()
 		if step.Ref != nil {
 			s.Ref = nil
+			getStepAction := GetStepActionFunc(tekton, tr.Namespace)
```

Comment:
Could you help me to understand the reason this needs to be in the loop?

Reply:
The follow up PR needs to have it in the loop to get the step.Ref.Resolver
https://github.com/tektoncd/pipeline/pull/7321